### PR TITLE
Plugins: Support > 1 levels of plugin dependencies

### PR DESCRIPTION
**Pull request by wbrowne**
**What is this feature?**
Adds support for installing 2 or more levels of plugin dependencies. For example the `redis-explorer-app` has a dependency on the `redis-app`. This is supported already today. However, `redis-app` has a dependency on `redis-datasource` (IE `redis-explorer-app` -> `redis-app` -> `redis-datasource`). This pull request ensures that we also install this plugins (and beyond if necessary).

Before:
```
INFO [07-08|11:06:38] Downloaded and extracted redis-explorer-app v2.1.1 zip successfully to /Users/wbrowne/dev/grafana/data/plugins/redis-explorer-app logger=installer.fs
INFO [07-08|11:06:38] Fetching redis-app dependencies...       logger=plugin.installer
INFO [07-08|11:06:38] Downloaded and extracted redis-app v2.2.1 zip successfully to /Users/wbrowne/dev/grafana/data/plugins/redis-app logger=installer.fs
DEBUG[07-08|11:06:38] Loading plugin source...                 logger=plugin.loader source=external
INFO [07-08|11:06:38] Plugin registered                        logger=plugins.registration pluginId=redis-explorer-app
INFO [07-08|11:06:38] Plugin registered                        logger=plugins.registration pluginId=redis-enterprise-software-datasource
INFO [07-08|11:06:38] Plugin registered                        logger=plugins.registration pluginId=redis-app
INFO [07-08|11:06:38] Plugin registered                        logger=plugins.registration pluginId=redis-latency-panel
INFO [07-08|11:06:38] Plugin registered                        logger=plugins.registration pluginId=redis-cli-panel
INFO [07-08|11:06:38] Plugin registered                        logger=plugins.registration pluginId=redis-cpu-panel
INFO [07-08|11:06:38] Plugin registered                        logger=plugins.registration pluginId=redis-gears-panel
INFO [07-08|11:06:38] Plugin registered                        logger=plugins.registration pluginId=redis-keys-panel
DEBUG[07-08|11:06:38] Plugin source loaded                     logger=plugin.loader source=external plugins="redis-app, redis-cli-panel, redis-cpu-panel, redis-enterprise-software-datasource, redis-explorer-app, redis-gears-panel, redis-keys-panel, redis-latency-panel" duration=16.727792ms
```

After:
```
INFO [07-08|11:04:25] Downloaded and extracted redis-explorer-app v2.1.1 zip successfully to /Users/wbrowne/dev/grafana/data/plugins/redis-explorer-app logger=installer.fs
INFO [07-08|11:04:25] Fetching redis-explorer-app dependency redis-app... logger=plugin.installer
INFO [07-08|11:04:26] Downloaded and extracted redis-app v2.2.1 zip successfully to /Users/wbrowne/dev/grafana/data/plugins/redis-app logger=installer.fs
INFO [07-08|11:04:26] Fetching redis-app dependency redis-datasource... logger=plugin.installer
INFO [07-08|11:04:31] Downloaded and extracted redis-datasource v2.1.1 zip successfully to /Users/wbrowne/dev/grafana/data/plugins/redis-datasource logger=installer.fs
DEBUG[07-08|11:04:31] Loading plugin source...                 logger=plugin.loader source=external
INFO [07-08|11:04:31] Plugin registered                        logger=plugins.registration pluginId=redis-datasource
DEBUG[07-08|11:04:31] Plugin source loaded                     logger=plugin.loader source=external plugins=redis-datasource duration=224.518125ms
DEBUG[07-08|11:04:31] Loading plugin source...                 logger=plugin.loader source=external
INFO [07-08|11:04:32] Plugin registered                        logger=plugins.registration pluginId=redis-app
INFO [07-08|11:04:32] Plugin registered                        logger=plugins.registration pluginId=redis-keys-panel
INFO [07-08|11:04:32] Plugin registered                        logger=plugins.registration pluginId=redis-latency-panel
INFO [07-08|11:04:32] Plugin registered                        logger=plugins.registration pluginId=redis-cli-panel
INFO [07-08|11:04:32] Plugin registered                        logger=plugins.registration pluginId=redis-cpu-panel
INFO [07-08|11:04:32] Plugin registered                        logger=plugins.registration pluginId=redis-gears-panel
DEBUG[07-08|11:04:32] Plugin source loaded                     logger=plugin.loader source=external plugins="redis-app, redis-cli-panel, redis-cpu-panel, redis-gears-panel, redis-keys-panel, redis-latency-panel" duration=13.8365ms
DEBUG[07-08|11:04:32] Loading plugin source...                 logger=plugin.loader source=external
INFO [07-08|11:04:32] Plugin registered                        logger=plugins.registration pluginId=redis-explorer-app
INFO [07-08|11:04:32] Plugin registered                        logger=plugins.registration pluginId=redis-enterprise-software-datasource
DEBUG[07-08|11:04:32] Plugin source loaded                     logger=plugin.loader source=external plugins="redis-enterprise-software-datasource, redis-explorer-app" duration=8.367375ms
```

**Why do we need this feature?**

While there is only one plugin (`redis-explorer-app`) that relies on this functionality - this should be supported for correctness.

**Who is this feature for?**

Plugin users / operators.

**Special notes for your reviewer:**
These changes are reflected in both the grafana CLI and the plugin catalog.

### CODE REVIEW

File: pkg/plugins/manager/installer.go [hunk 1]
```diff
@@ -54,89 +54,90 @@ func (m *PluginInstaller) Add(ctx context.Context, pluginID, version string, opt
...
+	for _, dep := range archive.Dependencies {
+		m.log.Info(fmt.Sprintf("Fetching %s dependency %s...", pluginID, dep.ID))
+
+		err = m.Add(ctx, dep.ID, dep.Version, opts)
+		if err != nil {
+			return fmt.Errorf("%v: %w", fmt.Sprintf("failed to download plugin %s from repository", dep.ID), err)
+		}
+	}
+
+	_, err = m.pluginLoader.Load(ctx, sources.NewLocalSource(plugins.ClassExternal, []string{archive.Path}))
```
Comment (andresmgot) [inline pkg/plugins/manager/installer.go:71]: curious: what's this check for?

**Review by andresmgot — APPROVED** [review]
LGTM

File: pkg/plugins/manager/installer.go (hunk 1 above)
Comment (wbrowne) [reply pkg/plugins/manager/installer.go:71]: This is the code that actually loads the new plugin(s) once they've been extracted to disk

**Review by marefr — APPROVED** [review]
LGTM. Is there a chance for circular dependencies that we need to cater for? 🤔

**Comment by wbrowne** [comment]
> LGTM. Is there a chance for circular dependencies that we need to cater for? 🤔

Yes 😉 that exists today. Will create an issue for it!

**Comment by marefr** [comment]
> > LGTM. Is there a chance for circular dependencies that we need to cater for? 🤔
> 
> Yes 😉 ~that exists today~. Will create an issue for it and do it in a follow up.

But today it's maximum 1 level right so can't see that happen? If not a big chance happening might be worth punt until someone experience it :)

File: pkg/plugins/manager/installer.go [hunk 2]
```diff
@@ -54,89 +56,98 @@ func (m *PluginInstaller) Add(ctx context.Context, pluginID, version string, opt
 		return err
 	}
 
+	if m.installing[pluginID] {
```
Comment (wbrowne) [inline pkg/plugins/manager/installer.go:59]: Maybe a lock necessary here or just change to `sync.Map`

**Review by marefr — APPROVED** [review]
LGTM

**Pull request merged**